2022-09-28-introduce-q-d.md

```diff
@@ -0,0 +1,241 @@
+
+## Background
+
+Testing stateful or rather effectful code using QuickCheck is not new: In particular, techniques to test _Monadic_ code with QuickCheck have been introduced in [Claessen & Hughes, 2002](https://dl.acm.org/doi/10.1145/636517.636527). `quickcheck-dynamic` is based on a state-machine approach originally implemented by Quviq in closed-source [Erlang version of QuickCheck](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.148.6554&rep=rep1&type=pdf) and put to use to test various systems as explained in John Hughes' [paper](https://publications.lib.chalmers.se/records/fulltext/232550/local_232550.pdf).
+
+IOG already has had experience with state-machine based testing in the Consensus [Storage layer](https://github.com/input-output-hk/ouroboros-network/blob/nfrisby/old-benchmark-ledger-ops/ouroboros-consensus-test/README.md#L1) using [quickcheck-state-machine](https://hackage.haskell.org/package/quickcheck-state-machine) library, but this was not widespread practice across the various teams.
+
+When IOG started working on the Plutus Smart Contracts language and application framework, Quviq's consultancy was sought after to help test the platform and build tools for future Smart Contract implementors. This effort lead to the development of a custom library for testing smart contracts which in turn leverages quickcheck-dynamic's state-machine model and dynamic logic language.
+
+As quickcheck-dynamic matured and became independent of the Plutus framework, it attracted interest from other teams willing to invest into model-based testing and reuse existing effort. This finally lead to publication of the library as an independent package on [Hackage](https://hackage.haskell.org/package/quickcheck-dynamic) in the hope it will be useful to a wider audience.
+
+## Use Cases
+
+### Example: Thread Registry
+
+The library comes with a complete example defining a model and reference implementation for a _Thread Registry_. It's inspired by a similar example in Erlang from a couple of papers:
+  * [How well are your requirements tested?](https://publications.lib.chalmers.se/records/fulltext/232552/local_232552.pdf)
+  * and [Understanding Formal Specifications through Good Examples](https://mengwangoxf.github.io/Papers/Erlang18.pdf)
+
+The tests here use IOG's concurrent execution simulator library [io-sim](https://github.com/input-output-hk/io-sim) to speed-up testing.
```

Comment:
TODO for @MaximilianAlgehed: rebase on the latest version of io-sim to get the merged (?) fix to `threadStatus` in the registry example and make sure the registry example actually passes the tests under io-sim.